core/src/main/java/hudson/model/LoadBalancer.java

```diff
@@ -84,8 +86,12 @@ public String hash(ExecutorChunk node) {
                         return node.getName();
                     }
                 });
-                for (ExecutorChunk ec : ws.works(i).applicableExecutorChunks())
-                    hash.add(ec,ec.size()*100);
+
+                Map<ExecutorChunk, Integer> toAdd = new HashMap<ExecutorChunk, Integer>();
```

Comment:
Can you drop a comment here explaining that this is a performance optimization so no one will be tempted to *simplify* the code back? Also, It should be possible to initialize the map with expected capacity.